File: backend/app/models/lab_result.py
"""
Lab results models for test orders and results.
"""

import uuid
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING, Any

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel

if TYPE_CHECKING:
    from app.models.doctor import Doctor
    from app.models.patient import Patient


class LabOrderStatus(str, Enum):
    """Lab order status types."""

    ORDERED = "ordered"
    SAMPLE_COLLECTED = "sample_collected"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    REJECTED = "rejected"


class LabTestPriority(str, Enum):
    """Lab test priority."""

    ROUTINE = "routine"
    URGENT = "urgent"
    STAT = "stat"  # Immediate


class LabResultStatus(str, Enum):
    """Lab result status."""

    PENDING = "pending"
    PRELIMINARY = "preliminary"
    FINAL = "final"
    CORRECTED = "corrected"
    AMENDED = "amended"


class LabOrder(BaseModel):
    """
    Lab order/requisition model.

    Attributes:
        patient_id: Patient for whom tests are ordered
        doctor_id: Doctor who ordered the tests
        appointment_id: Associated appointment (optional)
        order_date: When the order was placed
        tests_ordered: List of test names/codes requested
        priority: Order priority (routine, urgent, stat)
        status: Current order status
        clinical_notes: Clinical indication for tests
        diagnosis_codes: ICD-10 or similar codes
        sample_collected_at: When sample was collected
        expected_completion: Expected result date
        lab_provider: External lab name (if applicable)
        lab_order_id: External lab's order ID
    """

    __tablename__ = "lab_orders"

    # Relations
    patient_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("patients.id"),
        nullable=False,
        index=True,
    )
    doctor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("doctors.id"),
        nullable=False,
        index=True,
    )
    appointment_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("appointments.id"),
        nullable=True,
        index=True,
    )

    # Order details
    order_date: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )
    tests_ordered: Mapped[list[str]] = mapped_column(
        JSONB,
        nullable=False,
    )
    priority: Mapped[str] = mapped_column(
        String(20),
        default=LabTestPriority.ROUTINE.value,
    )
    status: Mapped[str] = mapped_column(
        String(20),
        default=LabOrderStatus.ORDERED.value,
        index=True,
    )

    # Clinical info
    clinical_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    diagnosis_codes: Mapped[list[str] | None] = mapped_column(JSONB, nullable=True)

    # Sample & timing
    sample_collected_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    expected_completion: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    # External lab integration
    lab_provider: Mapped[str | None] = mapped_column(String(100), nullable=True)
    lab_order_id: Mapped[str | None] = mapped_column(String(100), nullable=True, index=True)

    # Metadata
    metadata: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)

    # Relationships
    patient: Mapped["Patient"] = relationship("Patient", back_populates="lab_orders")
    doctor: Mapped["Doctor"] = relationship("Doctor", back_populates="lab_orders")
    results: Mapped[list["LabResult"]] = relationship(
        "LabResult",
        back_populates="order",
        cascade="all, delete-orphan",
    )
    reports: Mapped[list["LabReport"]] = relationship(
        "LabReport",
        back_populates="order",
        cascade="all, delete-orphan",
    )

    def __repr__(self) -> str:
        return f"<LabOrder {self.id} ({self.status})>"


class LabResult(BaseModel):
    """
    Individual lab test result.

    Attributes:
        order_id: Associated lab order
        test_name: Name of the test (e.g., "Hemoglobin")
        test_code: Standard code (LOINC, local code)
        value: Test result value
        value_numeric: Numeric value for trending
        unit: Unit of measurement
        reference_range_min: Lower bound of normal range
        reference_range_max: Upper bound of normal range
        reference_range_text: Textual reference (e.g., "Negative", "< 200")
        is_abnormal: Flag for abnormal values
        abnormal_flag: Type of abnormality (H=high, L=low, etc.)
        status: Result status
        result_date: When result was finalized
        notes: Interpretation notes
        performed_by: Lab technician/analyzer
    """

    __tablename__ = "lab_results"

    # Relations
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("lab_orders.id"),
        nullable=False,
        index=True,
    )

    # Test identification
    test_name: Mapped[str] = mapped_column(String(200), nullable=False, index=True)
    test_code: Mapped[str | None] = mapped_column(String(50), nullable=True)
    test_category: Mapped[str | None] = mapped_column(String(100), nullable=True)  # CBC, LFT, etc.

    # Result value
    value: Mapped[str] = mapped_column(Text, nullable=False)  # Can be text or numeric
    value_numeric: Mapped[float | None] = mapped_column(Float, nullable=True)  # For trending
    unit: Mapped[str | None] = mapped_column(String(50), nullable=True)

    # Reference ranges
    reference_range_min: Mapped[float | None] = mapped_column(Float, nullable=True)
    reference_range_max: Mapped[float | None] = mapped_column(Float, nullable=True)
    reference_range_text: Mapped[str | None] = mapped_column(String(200), nullable=True)

    # Abnormality flags
    is_abnormal: Mapped[bool] = mapped_column(Boolean, default=False, index=True)
    abnormal_flag: Mapped[str | None] = mapped_column(
        String(10), nullable=True
    )  # H, L, HH, LL, A

    # Status & timing
    status: Mapped[str] = mapped_column(
        String(20),
        default=LabResultStatus.FINAL.value,
    )
    result_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    # Additional info
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    performed_by: Mapped[str | None] = mapped_column(String(100), nullable=True)
    methodology: Mapped[str | None] = mapped_column(String(100), nullable=True)

    # Metadata
    metadata: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)

    # Relationships
    order: Mapped["LabOrder"] = relationship("LabOrder", back_populates="results")

    @property
    def is_high(self) -> bool:
        """Check if value is above normal range."""
        if self.value_numeric is None or self.reference_range_max is None:
            return False
        return self.value_numeric > self.reference_range_max

    @property
    def is_low(self) -> bool:
        """Check if value is below normal range."""
        if self.value_numeric is None or self.reference_range_min is None:
            return False
        return self.value_numeric < self.reference_range_min

    def __repr__(self) -> str:
        return f"<LabResult {self.test_name}: {self.value} {self.unit or ''}>"


class LabReport(BaseModel):
    """
    Lab report document (PDF, HL7, etc.).

    Attributes:
        order_id: Associated lab order
        report_type: Type of report (pdf, hl7, text)
        file_path: Path to stored file
        file_url: External URL (if applicable)
        file_size: File size in bytes
        parsed_data: Extracted structured data
        raw_content: Raw text content
        received_at: When report was received
    """

    __tablename__ = "lab_reports"

    # Relations
    order_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("lab_orders.id"),
        nullable=False,
        index=True,
    )

    # Report details
    report_type: Mapped[str] = mapped_column(String(20), nullable=False)  # pdf, hl7, text
    file_path: Mapped[str | None] = mapped_column(Text, nullable=True)
    file_url: Mapped[str | None] = mapped_column(Text, nullable=True)
    file_size: Mapped[int | None] = mapped_column(Integer, nullable=True)
    mime_type: Mapped[str | None] = mapped_column(String(100), nullable=True)

    # Parsed content
    parsed_data: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    raw_content: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Timing
    received_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
    )

    # Relationships
    order: Mapped["LabOrder"] = relationship("LabOrder", back_populates="reports")

    def __repr__(self) -> str:
        return f"<LabReport {self.id} ({self.report_type})>"

File: mobile/lib/core/models/lab_result.dart
/// Lab result models

/// Lab order model
class LabOrder {
  final String id;
  final String patientId;
  final String doctorId;
  final String? appointmentId;
  final DateTime orderDate;
  final List<String> testsOrdered;
  final String priority;
  final String status;
  final String? clinicalNotes;
  final List<String>? diagnosisCodes;
  final DateTime? sampleCollectedAt;
  final DateTime? expectedCompletion;
  final String? labProvider;
  final String? labOrderId;
  final Map<String, dynamic>? metadata;

  // Enriched data
  final String? patientName;
  final String? doctorName;
  final int resultsCount;
  final int abnormalCount;

  final DateTime createdAt;
  final DateTime updatedAt;

  LabOrder({
    required this.id,
    required this.patientId,
    required this.doctorId,
    this.appointmentId,
    required this.orderDate,
    required this.testsOrdered,
    required this.priority,
    required this.status,
    this.clinicalNotes,
    this.diagnosisCodes,
    this.sampleCollectedAt,
    this.expectedCompletion,
    this.labProvider,
    this.labOrderId,
    this.metadata,
    this.patientName,
    this.doctorName,
    this.resultsCount = 0,
    this.abnormalCount = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory LabOrder.fromJson(Map<String, dynamic> json) {
    return LabOrder(
      id: json['id'],
      patientId: json['patient_id'],
      doctorId: json['doctor_id'],
      appointmentId: json['appointment_id'],
      orderDate: DateTime.parse(json['order_date']),
      testsOrdered: List<String>.from(json['tests_ordered'] ?? []),
      priority: json['priority'] ?? 'routine',
      status: json['status'] ?? 'ordered',
      clinicalNotes: json['clinical_notes'],
      diagnosisCodes: json['diagnosis_codes'] != null
          ? List<String>.from(json['diagnosis_codes'])
          : null,
      sampleCollectedAt: json['sample_collected_at'] != null
          ? DateTime.parse(json['sample_collected_at'])
          : null,
      expectedCompletion: json['expected_completion'] != null
          ? DateTime.parse(json['expected_completion'])
          : null,
      labProvider: json['lab_provider'],
      labOrderId: json['lab_order_id'],
      metadata: json['metadata'],
      patientName: json['patient_name'],
      doctorName: json['doctor_name'],
      resultsCount: json['results_count'] ?? 0,
      abnormalCount: json['abnormal_count'] ?? 0,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'patient_id': patientId,
      'doctor_id': doctorId,
      'appointment_id': appointmentId,
      'tests_ordered': testsOrdered,
      'priority': priority,
      'clinical_notes': clinicalNotes,
      'diagnosis_codes': diagnosisCodes,
      'expected_completion': expectedCompletion?.toIso8601String(),
      'lab_provider': labProvider,
    };
  }

  String get statusDisplay {
    switch (status) {
      case 'ordered':
        return 'Ordered';
      case 'sample_collected':
        return 'Sample Collected';
      case 'in_progress':
        return 'In Progress';
      case 'completed':
        return 'Completed';
      case 'cancelled':
        return 'Cancelled';
      default:
        return status;
    }
  }

  String get priorityDisplay {
    switch (priority) {
      case 'routine':
        return 'Routine';
      case 'urgent':
        return 'Urgent';
      case 'stat':
        return 'STAT';
      default:
        return priority;
    }
  }

  bool get isCompleted => status == 'completed';
  bool get isPending => status != 'completed' && status != 'cancelled';
  bool get hasAbnormalResults => abnormalCount > 0;
}

/// Lab result model
class LabResult {
  final String id;
  final String orderId;
  final String testName;
  final String? testCode;
  final String? testCategory;
  final String value;
  final double? valueNumeric;
  final String? unit;
  final double? referenceRangeMin;
  final double? referenceRangeMax;
  final String? referenceRangeText;
  final bool isAbnormal;
  final String? abnormalFlag;
  final String status;
  final DateTime? resultDate;
  final String? notes;
  final String? performedBy;
  final String? methodology;
  final DateTime createdAt;
  final DateTime updatedAt;

  LabResult({
    required this.id,
    required this.orderId,
    required this.testName,
    this.testCode,
    this.testCategory,
    required this.value,
    this.valueNumeric,
    this.unit,
    this.referenceRangeMin,
    this.referenceRangeMax,
    this.referenceRangeText,
    required this.isAbnormal,
    this.abnormalFlag,
    required this.status,
    this.resultDate,
    this.notes,
    this.performedBy,
    this.methodology,
    required this.createdAt,
    required this.updatedAt,
  });

  factory LabResult.fromJson(Map<String, dynamic> json) {
    return LabResult(
      id: json['id'],
      orderId: json['order_id'],
      testName: json['test_name'],
      testCode: json['test_code'],
      testCategory: json['test_category'],
      value: json['value'],
      valueNumeric: json['value_numeric']?.toDouble(),
      unit: json['unit'],
      referenceRangeMin: json['reference_range_min']?.toDouble(),
      referenceRangeMax: json['reference_range_max']?.toDouble(),
      referenceRangeText: json['reference_range_text'],
      isAbnormal: json['is_abnormal'] ?? false,
      abnormalFlag: json['abnormal_flag'],
      status: json['status'] ?? 'final',
      resultDate: json['result_date'] != null
          ? DateTime.parse(json['result_date'])
          : null,
      notes: json['notes'],
      performedBy: json['performed_by'],
      methodology: json['methodology'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  String get displayValue {
    if (unit != null) {
      return '$value $unit';
    }
    return value;
  }

  String get referenceRangeDisplay {
    if (referenceRangeText != null && referenceRangeText!.isNotEmpty) {
      return referenceRangeText!;
    }
    if (referenceRangeMin != null && referenceRangeMax != null) {
      return '$referenceRangeMin - $referenceRangeMax${unit != null ? ' $unit' : ''}';
    }
    if (referenceRangeMax != null) {
      return '< $referenceRangeMax${unit != null ? ' $unit' : ''}';
    }
    if (referenceRangeMin != null) {
      return '> $referenceRangeMin${unit != null ? ' $unit' : ''}';
    }
    return 'N/A';
  }

  bool get isHigh => abnormalFlag == 'H' || abnormalFlag == 'HH';
  bool get isLow => abnormalFlag == 'L' || abnormalFlag == 'LL';
  bool get isCritical => abnormalFlag == 'HH' || abnormalFlag == 'LL';
}

/// Lab result trend data point
class LabResultTrendPoint {
  final DateTime date;
  final double value;
  final bool isAbnormal;

  LabResultTrendPoint({
    required this.date,
    required this.value,
    required this.isAbnormal,
  });

  factory LabResultTrendPoint.fromJson(Map<String, dynamic> json) {
    return LabResultTrendPoint(
      date: DateTime.parse(json['date']),
      value: json['value'].toDouble(),
      isAbnormal: json['is_abnormal'] ?? false,
    );
  }
}

/// Lab result trend response
class LabResultTrend {
  final String testName;
  final String? unit;
  final double? referenceRangeMin;
  final double? referenceRangeMax;
  final List<LabResultTrendPoint> dataPoints;

  LabResultTrend({
    required this.testName,
    this.unit,
    this.referenceRangeMin,
    this.referenceRangeMax,
    required this.dataPoints,
  });

  factory LabResultTrend.fromJson(Map<String, dynamic> json) {
    return LabResultTrend(
      testName: json['test_name'],
      unit: json['unit'],
      referenceRangeMin: json['reference_range_min']?.toDouble(),
      referenceRangeMax: json['reference_range_max']?.toDouble(),
      dataPoints: (json['data_points'] as List)
          .map((point) => LabResultTrendPoint.fromJson(point))
          .toList(),
    );
  }
}

/// Lab report model
class LabReport {
  final String id;
  final String orderId;
  final String reportType;
  final String? filePath;
  final String? fileUrl;
  final int? fileSize;
  final String? mimeType;
  final Map<String, dynamic>? parsedData;
  final DateTime receivedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  LabReport({
    required this.id,
    required this.orderId,
    required this.reportType,
    this.filePath,
    this.fileUrl,
    this.fileSize,
    this.mimeType,
    this.parsedData,
    required this.receivedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory LabReport.fromJson(Map<String, dynamic> json) {
    return LabReport(
      id: json['id'],
      orderId: json['order_id'],
      reportType: json['report_type'],
      filePath: json['file_path'],
      fileUrl: json['file_url'],
      fileSize: json['file_size'],
      mimeType: json['mime_type'],
      parsedData: json['parsed_data'],
      receivedAt: DateTime.parse(json['received_at']),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }
}

/// Lab statistics model
class LabStatistics {
  final int totalOrders;
  final int pendingOrders;
  final int completedOrders;
  final int totalResults;
  final int abnormalResults;
  final List<Map<String, dynamic>> mostOrderedTests;
  final double? averageTurnaroundHours;

  LabStatistics({
    required this.totalOrders,
    required this.pendingOrders,
    required this.completedOrders,
    required this.totalResults,
    required this.abnormalResults,
    required this.mostOrderedTests,
    this.averageTurnaroundHours,
  });

  factory LabStatistics.fromJson(Map<String, dynamic> json) {
    return LabStatistics(
      totalOrders: json['total_orders'] ?? 0,
      pendingOrders: json['pending_orders'] ?? 0,
      completedOrders: json['completed_orders'] ?? 0,
      totalResults: json['total_results'] ?? 0,
      abnormalResults: json['abnormal_results'] ?? 0,
      mostOrderedTests: List<Map<String, dynamic>>.from(
        json['most_ordered_tests'] ?? [],
      ),
      averageTurnaroundHours: json['average_turnaround_hours']?.toDouble(),
    );
  }
}
